orttraining/orttraining/core/graph/training_op_defs.cc

```diff
@@ -142,13 +142,189 @@ TensorProto ToDimensionOneFloatTensor(float value) {
+                      grad = Div (adj_weight_BD, sum_weights)
+                      d_loss = Mul (grad, dY)
+                  )");
+      } else {
+        builder.Add("d_loss = Mul (adj_weight_BD, dY)");
+      }
+    } else {
+      builder.Add(R"(
+                    not_ignored_BD = Not (ignored_BD)
+                    adj_weight_BD = CastLike (not_ignored_BD, dY)
+                )");
+      if (mean_reduction) {
+        builder.Add(R"(
+                      sum_weights = ReduceSum <keepdims = 0> (adj_weight_BD)
+                      grad = Div (adj_weight_BD, sum_weights)
+                      d_loss = Mul (grad, dY)
+                  )");
+      } else {
+        builder.Add("d_loss = Mul (adj_weight_BD, dY)");
+      }
+    }
+  } else {
+    if (has_weight) {
+      builder.Add("elt_weight = Gather (weight, label)");
+      if (mean_reduction) {
+        // backward-prop for y = ReduceSum (loss * elt_weight) / ReduceSum(elt_weight)
+        builder.Add(R"(
+                      sum_weights = ReduceSum <keepdims = 0> (elt_weight)
+                      grad = Div (elt_weight, sum_weights)
+                      d_loss = Mul(grad, dY)
+                  )");
+      } else {
+        // common backward-prop for y = ReduceSum(loss * elt_weight) and y = loss * elt_weight
+        builder.Add("d_loss = Mul(elt_weight, dY)");
+      }
+    } else {
+      if (mean_reduction) {
+        // backward-prop for y = ReduceSum (loss) / Size(label)
+        builder.Add(R"(
+                      count = Size(label)
+                      count_T = CastLike (count, dY)
+                      d_div = Div (dY, count_T)
+                      BD = Shape (label)
+                      d_loss = Expand (d_div, BD)
+                  )");
+      } else {
+        // common backward-prop for y = ReduceSum(loss) and y = loss
+        builder.Add(R"(
+                      BD = Shape (label)
+                      d_loss = Expand (dY, BD)
+                  )");
+      }
+    }
+  }
+
+  // Step 2: Compute d_logits from d_loss
+  // The gradient is essentially "probability - (1 if true-label else 0)", complicated
+  // by the reshaping for the general case.
+  builder.Add(R"(
+                d_loss_B1Dopt = Unsqueeze (d_loss, axes1)
+                reshape_arg = Constant < value = int64[3] {0, 0, -1} > ()
+                d_loss_B1D = Reshape (d_loss_B1Dopt, reshape_arg)
+                orig_shape = Shape (log_prob)
+                log_prob_BCD = Reshape (log_prob, reshape_arg)
+                prob_BCD = Exp (log_prob_BCD)
+
+                label_BD = Flatten (label) # convert from [B, d1, d2, ...] to [B, D = d1 * d2 * ...]
+
+                zero_one = Constant < value = int32[2] {0, 1}>()
+                zero_one_typed = CastLike (zero_one, prob_BCD)
+                C1d = Shape <start = 1, end = 2> (prob_BCD)
+                C = Squeeze(C1d)
+                one_hot_label_BCD = OneHot <axis=1> (label_BD, C, zero_one_typed)
```

Comment:
>OneHot [](http://example.com/codeflow?start=36&length=6)

Is there a less scary alternative to this that you can think of?

Reply:
Not really. There is an alternative encoding which is a bit longer, and which will probably make efficient compilation harder, so not sure if would be helpful. What metric would you use to choose one as being better (or scarier) than another?